### PR TITLE
Simplify specifying torch dependency

### PR DESCRIPTION
**Pull request by kousu**
## Checklist

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [ ] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [ ] I've added relevant tests for my contribution
- [ ] I've updated the documentation and/or added correct docstrings
- [ ] I've assigned a reviewer
- [ ] I've consulted [ADS's internal developer documentation](https://github.com/neuropoly/axondeepseg/wiki/Guidelines-for-contribution) to ensure my contribution is in line with any relevant design decisions

## Description

~~Using --extra-index-url instead of --find-links  -- which means we can trust ivadomed to specify the version it needs instead.~~

ivadomed [now ](https://github.com/ivadomed/ivadomed/pull/1129) specifies `torch` as a standard dependency, without giving special install instructions just to handle it, which means there's no need to pin a torch version here in axondeepseg either.

Simultaneously, move the pinning of `torch` to its CPU build into an optional step in _our_ install instructions. The default `torch` build works everywhere, its only problem is that it's large and useless for people who don't own GPUs. But for people who do own GPUs, ADS can make use of them, so we want to support both. Doing things this way brings the install in line with what [torch themselves wants us to do](https://pytorch.org/get-started/locally/)

## Linked issues

Fixes(?) https://github.com/ivadomed/ivadomed/issues/861#issuecomment-885957932

Related: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3790

### CODE REVIEW

**Comment by hermancollin** [comment]
I tested it and as far as I know everything works as expected. On linux, it downloads the cpu build. 
```
Collecting torch~=1.8.1
  Using cached https://download.pytorch.org/whl/cpu/torch-1.8.1%2Bcpu-cp38-cp38-linux_x86_64.whl (169.1 MB)
Collecting torchvision~=0.9.1
  Using cached https://download.pytorch.org/whl/cpu/torchvision-0.9.1%2Bcpu-cp38-cp38-linux_x86_64.whl (13.3 MB)

```
IIUC, to enable GPU support, you only need to change the `--extra-index-url` to https://download.pytorch.org/whl/cu113? When I try it, the linux torch-gpu *big boy* is downloaded:

```
Collecting torch~=1.8.1
  Downloading torch-1.8.1-cp38-cp38-manylinux1_x86_64.whl (804.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 804.1/804.1 MB 1.7 MB/s eta 0:00:00
Collecting torchvision~=0.9.1
  Downloading torchvision-0.9.1-cp38-cp38-manylinux1_x86_64.whl (17.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.4/17.4 MB 9.5 MB/s eta 0:00:00
```

Looks like everything works fine. (Weirdly enough, this GPU build looks faster than what I was used to on bireli... maybe I'm crazy). I understand that GPU support on windows will require https://download.pytorch.org/whl/cu102 as the `--extra-index-url` link, is that correct? Nonetheless, I don't recall a lot of situations where windows users needed GPU support.

From a first glance, it looks like this contribution solves the torch dependency problem for downstream packages, or at least makes it way easier to manage. Thank you @kousu and @kanishk16 for figuring this out.

File: environment.yml
Comment (hermancollin) [inline environment.yml:34]:
```suggestion
    - ivadomed==2.9.6
```
(to remember to change this after the new release)

**Comment by kousu** [comment]
> IIUC, to enable GPU support, you only need to change the `--extra-index-url` to https://download.pytorch.org/whl/cu113? When I try it, the linux torch-gpu _big boy_ is downloaded:
> 
> I understand that GPU support on windows will require https://download.pytorch.org/whl/cu102 as the `--extra-index-url` link, is that correct? Nonetheless, I don't recall a lot of situations where windows users needed GPU support.

I'm [going to recommend](https://github.com/ivadomed/ivadomed/pull/1125#issuecomment-1139752602) that you just remove `--extra-index-url` entirely to enable GPU support.

torch has make a couple quirky decisions: the GPU builds come with CUDA (unlike tensorflow, which expects the sysadmin to install CUDA systemwide), and on Linux, GPU builds are the default, while on Windows, they're not. This is my [theory why](https://github.com/ivadomed/ivadomed/pull/1129#issuecomment-1139730419):

> I think they must assume if you're using Linux you're hardcore and/or rich and/or using a data center, where you can afford to waste 700MB if you don't have GPUs, but if you do have GPUs then the time spent to remember how to add `--extra-index-url` .... costs more than just wasting the 700MB, while on the other hand and if you're using Windows you're probably a newbie and you need to be hand-held. torch is Facebook's project, designed for Facebook Research, and their researchers are probably using Windows or macOS locally, without GPUs, but deploying on their internal data centers, which have both GPUs and very fast internet connections.

The reason `cu113` is in the air is that it's a corruption of [`cu111`](https://github.com/ivadomed/ivadomed/pull/1129/files#diff-8dd5d07318f6b61b9015b03192e28c53734e34b06a1c3b589685dd8c0d249dc9L6) which was because `romane.neuro.polymtl.ca` is too new: it has [compute capability](https://developer.nvidia.com/cuda-gpus#compute) 8.6 but the [torch on pypi](https://pypi.org/project/torch/1.8.1/#copy-hash-modal-617ee84a-c35b-4262-905b-97b94994a98b) comes with CUDA 10.2. which only goes up to compute capability 7.0.  But really `romane` is an outlier and we shouldn't have tried to solve that in our software; I've solved that [a better way](https://github.com/neuropoly/computers/pull/367): its `/etc/pip.conf` now commands every python env on the system to use the CUDA 11 builds:

```
# romane:/etc/pip.conf
[install]
extra-index-url =
  https://download.pytorch.org/whl/cu116
  https://download.pytorch.org/whl/cu115
  https://download.pytorch.org/whl/cu114
  https://download.pytorch.org/whl/cu113
  https://download.pytorch.org/whl/cu112
  https://download.pytorch.org/whl/cu111
  https://download.pytorch.org/whl/cu110
```

It tells them to check all the sub-repos, to make sure to be backwards and forwards compatible with as many torch versions as possible. It's unfortunate they don't just have a single https://download.pytorch.org/whl/cuda or https://download.pytorch.org/whl/cu or https://download.pytorch.org/whl/gpu repo but :shrug: 

So simply removing `--extra-index-url` maintains the assumption that "Linux has GPU support" everywhere.

And eventually torch will standardize on CUDA 11 as the default for Linux, and we can drop the workaround.

For Windows, you can actually use the same pip.conf! Because there's CUDA 11 builds for Windows too, [e.g.](https://download.pytorch.org/whl/cu111/torch)

<details>

Screenshot 2022-05-31 at 11-18-49 https __download pytorch org

</details>

But Windows sysadmins could also meaningfully choose CUDA 10:

```
[install]
extra-index-url =
  https://download.pytorch.org/whl/cu102
  https://download.pytorch.org/whl/cu101
  https://download.pytorch.org/whl/cu100
```

since the Windows pypi.org builds don't have CUDA. It _would also_ work for Linux, it's simply redundant since these repos contain identical `*linux*.whl` files as pypi.org

And either situation could also do

```
[install]
extra-index-url =
  https://download.pytorch.org/whl/cpu
```

to force using the smaller CPU builds.

In any case, I think assuming 99% of people will be satisfied with the defaults, and then maybe for the rare exceptions we can help **sysadmins** set up /etc/pip.conf (C:\ProgramData\pip\pip.ini on Windows). We should probably put this key trick in the wiki, but not the docs: I want to get ivadomed out of the business of trying to teach people about the sprawling CUDA stack and I think you shouldn't be in it either, they just want their analyses to work.

> When I try it, the linux torch-gpu big boy is downloaded:
> 
> ```
> Collecting torch~=1.8.1
>   Downloading torch-1.8.1-cp38-cp38-manylinux1_x86_64.whl (804.1 MB)
>      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 804.1/804.1 MB 1.7 MB/s eta 0:00:00
> Collecting torchvision~=0.9.1
>   Downloading torchvision-0.9.1-cp38-cp38-manylinux1_x86_64.whl (17.4 MB)
>      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.4/17.4 MB 9.5 MB/s eta 0:00:00
> ```
> 
> Looks like everything works fine. (Weirdly enough, this GPU build looks faster than what I was used to on bireli... maybe I'm crazy)

Oh interesting! That's because you're not using the `cu113` build at all: there's no torch-1.8.1 in https://download.pytorch.org/whl/cu113/torch. Effectively, that was the same as removing `--extra-index-url`.

This is instead the [pypi version](https://pypi.org/project/torch/1.8.1/#copy-hash-modal-617ee84a-c35b-4262-905b-97b94994a98b) with CUDA 10. You should be able to confirm this with:

```
(ads_venv) p115628@bireli:~/axondeepseg$ python -c 'import torch; print(torch.version.cuda)'
10.2
```

If for some reason you wanted to try the CUDA 11 build you would actually have to use `--extra-index-url https://download.pytorch.org/whl/cu111`.

> From a first glance, it looks like this contribution solves the torch dependency problem for downstream packages, or at least makes it way easier to manage. Thank you @kousu and @kanishk16 for figuring this out.

Wonderful! Finally we can simplify this issue away.

**Comment by hermancollin** [comment]
On `bireli`, launching our quick integrity test with CUDA 11.1:
```
real    0m16.210s
user    0m52.484s
sys     0m3.251s
```
With CUDA 10.2:
```
real    0m9.065s
user    0m6.337s
sys     0m1.478s
```
!!!!!!

**Comment by hermancollin** [comment]
> Crazy! That suddenly makes it very clear to me why cuda11 isn't the default yet. What about `romane`? Is that equally slow?

On `romane` with CUDA 11.1
```
real    0m8.021s
user    0m12.308s
sys     0m2.939s
```
so, no it's not as slow. I'm guessing this is because `bireli`'s gpus are older. It seems a lot of people reported slower inference time with CUDA 11 in comparison with 10.2: https://github.com/pytorch/pytorch/issues/47908

CUDA 11.1 is shipped with cudnn 8 whereas CUDA 10.2 uses an older cudnn 7. It could be related to the cudnn version and if `torch.backends.cudnn.benchmark == True`, which seems to be the case for ivadomed, (e.g. [here](https://github.com/ivadomed/ivadomed/blob/7b76bf81a025cde3096fd1d686d6f3c0b8ce8f02/ivadomed/main.py#L28)) according to [this issue](https://github.com/pytorch/pytorch/issues/51044)

**Comment by kousu** [comment]
> It could be related to the cudnn version and if `torch.backends.cudnn.benchmark == True`, which seems to be the case for ivadomed, (e.g. [here](https://github.com/ivadomed/ivadomed/blob/7b76bf81a025cde3096fd1d686d6f3c0b8ce8f02/ivadomed/main.py#L28)) according to [this issue](https://github.com/pytorch/pytorch/issues/51044)

This is some pretty solid research. Good job digging those up. Let's raise a follow up with ivadomed to ask them if they could disable benchmarking by default, but that's not a blocker.

It looks like the ivadomed patch underneath is ready to go.

**Comment by kanishk16** [comment]
> On `bireli`, launching our quick integrity test with CUDA 11.1:
> 
> ```
> real    0m16.210s
> user    0m52.484s
> sys     0m3.251s
> ```
> 
> With CUDA 10.2:
> 
> ```
> real    0m9.065s
> user    0m6.337s
> sys     0m1.478s
> ```
> 
> !!!!!! (this test performs inference on an image) I think the time spent in the NVIDIA driver is halved

@hermancollin Is it possible for you to set `torch.backends.cudnn.benchmark = False` and then report the inference results?

**Comment by kanishk16** [comment]
> CUDA 11.1 is shipped with cudnn 8 whereas CUDA 10.2 is used from an older cudnn 7. It could be related to the cudnn version and if `torch.backends.cudnn.benchmark == True`, which seems to be the case for ivadomed, (e.g. [here](https://github.com/ivadomed/ivadomed/blob/7b76bf81a025cde3096fd1d686d6f3c0b8ce8f02/ivadomed/main.py#L28)) according to [this issue](https://github.com/pytorch/pytorch/issues/51044)

As per [official docs](https://pytorch.org/docs/stable/backends.html#torch.backends.cudnn.benchmark) and AFAIU (after referring to the [official discussion forum](https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936)) it may or may not hamper the time as per the situation. Since there are mixed opinions regarding the same, it would be great if could find the best tradeoff for us.

On a side note, we've already had discussions regarding exposing such parameters to the user https://github.com/ivadomed/ivadomed/issues/1092 and this would guide us in selecting the default parameters. But, this might be on a slightly low priority.

**Comment by hermancollin** [comment]
@kanishk16 I opened an issue to discuss this https://github.com/ivadomed/ivadomed/issues/1141

I tried setting benchmark to false but wasn't able to observe any difference.

**Comment by hermancollin** [comment]
Need to update the doc wrt. https://github.com/ivadomed/ivadomed/pull/1125

**Comment by kousu** [comment]
I'll try to move https://github.com/ivadomed/ivadomed/pull/1125 along, so that this PR can also be moved along.

**Comment by kousu** [comment]
For the record, this is me moving the docs along: https://github.com/ivadomed/ivadomed/pull/1184. But it seems like maybe it's a back-burner priority for them...

So I'm going to patch the docs for you too, and you should be able to merge that even without ivadomed getting time allocated. I'll going to remove the direct links to ivadomed so that those links can't break and instead explain in loose detail what people need to know as your users. Which isn't much. The whole point of this saga has been to streamline the situatuion so people **don't** need to overthink the installation.

File: environment.yml
Comment (kousu) [reply environment.yml]: I dropped it and instead documented the pip.conf path for people who want a smaller download.

**Comment by kousu** [comment]
Okay I've updated your docs: https://axondeepseg--642.org.readthedocs.build/en/642/documentation.html#installation

though apparently I don't know how to embed a code block inside a `.. note::`. Anyone know how to do that off the top of their heads?

**Comment by mathieuboudreau** [comment]
> though apparently I don't know how to embed a code block inside a `.. note::`. Anyone know how to do that off the top of their heads?

Done in 6430f85

RTD: https://axondeepseg.readthedocs.io/en/ng-unpin-torch/documentation.html#install-axondeepseg

Screenshot:

Screen Shot 2022-07-22 at 9 53 09 PM

**Review by hermancollin — APPROVED** [review]
@kousu thanks for taking care of the doc. I think these installation instructions are very clear. Maybe someone else would like to try the slightly different installation before merging, but this is good to go as far as I'm concerned.

**Comment by mathieuboudreau** [comment]
I'm testing this right now locally and on `bireli`, seems to be working fine so far, I'll merge if all goes well.

**Comment by kousu** [comment]
Thanks @mathieuboudreau. I was just writing up a final commit message myself. Here's my draft so far:

----

```
ivadomed now <https://github.com/ivadomed/ivadomed/pull/1129/> specifies torch
as a standard dependency, without giving special manual instructions for it.

This means there's finally no need to pin a torch version here in axondeepseg;
previously, we had to match their version in our environment.yml because it was
a manual step in their install instructions. But now the metadata makes pip handle it.

ivadomed gave special install instructions to try to handle the complicated mess that it
hardware (i.e. CPU vs GPU vs hyper modern CUDA-11-only GPUs) compatibility.
But now they're just leaning on the 80%-20% rule of thumb: >80% of people are covered
by the default torch builds. And we know how to handle the <20% cases of people with
CUDA-11-only GPUs or Windows machines with GPUs by having them install with
'--extra-index-url = https://download.pytorch.org/whl/<hardware-specific-build>'

axondeepseg was pushing the CPU version of torch on people, assuming that most users
would not own GPUs and that in the Linux case it would be very wasteful, maybe even
make ADS uninstallable for some users. We gave our own manual instructions that
duplicated ivadomed's for how to switch to the GPU version. That process meant double
the downloads and likely mismatched versions.

Instead, follow torch's defaults, like ivadomed now does. But since the users with
Linux and CPU-only hardware is fairly common and would be wasteful, highlight
--extra-index-url as an optional *pre*step, avoiding the double-download and
versioning mistakes.

Overall:

- Windows and macOS continue to have no GPU support
- Linux always has GPU support, even if it doesn't need it; this balloons the
  default install, but anyone using torch now has this same problem and the
  consensus seems to be just to pay for bigger disks that can tolerate the waste
- The relatively common case of Linux users without GPUs have an escape
  hatch to avoid the large download 
- Windows users with GPUs have a way to use them
  (`--extra-index-url https://download.pytorch.org/whl/cu102`), but we don't
  document nor support this
- Linux/Windows users with super new CUDA-11-only GPUs also have a
  way to use them, which we we also don't document nor support directly

And no one gets distracted with too many questions.

Fixes https://github.com/ivadomed/ivadomed/issues/861#issuecomment-885957932

Related: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3790

In passing, remove the redundant parts of the `conda env create` line.

Co-authored-by: Nick Guenther <nick.guenther@polymtl.ca>
Co-authored-by: Mathieu Boudreau, PhD <emb6150@gmail.com>
Co-authored-by: Armand <83031821+hermancollin@users.noreply.github.com>
```

**Review by mathieuboudreau — APPROVED** [review]
The new install worked well on CPU (my mac) and GPU (`bireli`), per passing testing suite.

@kousu your final commit message looks good to go for me, so this PR is ready to merge ASAP!

**Pull request merged**
